### PR TITLE
feat: Inpaint & Outpaint Improvements

### PR DESCRIPTION
**Pull request by blessedcoolant**
## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Optimization

## Have you discussed this change with the InvokeAI team?
- [x] Yes

## Description

# Coherence Mode

A new parameter called Coherence Mode has been added to Coherence Pass settings. This parameter controls what kind of Coherence Pass is done after Inpainting and Outpainting.

- Unmasked: This performs a complete unmasked image to image pass on the entire generation.
- Mask: This performs a masked image to image pass using your input mask as the coherence mask.
- Mask Edge [DEFAULT] - This performs as masked image to image pass on the edges of your mask to try and clear out the seams.

# Why The Coherence Masked Modes?

One of the issues with unmasked coherence pass arises when the diffusion process is trying to align detailed or organic objects. Because Image to Image tends change the image a little bit even at lower strengths, this ends up in the paste back process being slightly misaligned. By providing the mask to the Coherence Pass, we can try to eliminate this in those cases. While it will be impossible to address this for every image out there, having these options will allow the user to automate a lot of this. For everything else there's manual paint over with inpaint. 

# Graph Improvements

The graphs have now been refined quite a bit. We no longer do manual blurring of the masks anymore for outpainting. This is no longer needed because we now dilate the mask depending on the blur size while pasting back. As a result we got rid of quite a few nodes that were handling this in the older graph.

The graphs are also a lot cleaner now because we now tackle Scaled Dimensions & Coherence Mode completely independently.

Inpainting result seem very promising especially with the Mask Edge mode.

---

# New Infill Methods [Experimental]

We are currently trying out various new infill methods to see which ones might perform the best in outpainting. We may keep all of them or keep none. This will be decided as we test more.

## LaMa Infill

- Renabled LaMA infill in the UI.
- We are trying to get this to work without a memory overhead.

In order to use LaMa, you need to manually download and place the LaMa JIT model in `models/core/misc/lama/lama.pt`. You can download the JIT model from Sanster [here](https://github.com/Sanster/models/releases/download/add_big_lama/big-lama.pt) and rename it to `lama.pt` or you can use the script in the original LaMA repo to convert the base model to a JIT model yourself.

## CV2 Infill

- Added a new infilling method using CV2's Inpaint.

## Patchmatch Rescaling

Patchmatch infill input image is now downscaled and infilled. Patchmatch can be really slow at large resolutions and this is a pretty decent way to get around that. Additionally, downscaling might also provide a better patch match by avoiding larger areas to be infilled with repeating patches. But that's just the theory. Still testing it out.

## [optional] Are there any post deployment tasks we need to perform?

- If we decide to keep LaMA infill, then we will need to host the model and update the installer to download it as a core model.

### CODE REVIEW

**Comment by psychedelicious** [comment]
> The graphs are also a lot cleaner now because we now tackle Scaled Dimensions & Coherence Mode completely independently. 

The diff of this PR disagrees :P

**Comment by blessedcoolant** [comment]
> > The graphs are also a lot cleaner now because we now tackle Scaled Dimensions & Coherence Mode completely independently.
> 
> The diff of this PR disagrees :P

Haha. :D At least cleaner in my head. Coz now I don't get lost in them which would be a shame.

File: invokeai/frontend/web/src/features/parameters/store/generationSlice.ts [hunk 1]
```diff
@@ -87,7 +91,8 @@ export const initialGenerationState: GenerationState = {
   shouldUseNoiseSettings: false,
   steps: 50,
   threshold: 0,
-  tileSize: 32,
+  infillTileSize: 32,
+  infillPatchmatchDownscaleSize: 2,
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/parameters/store/generationSlice.ts:95]: This yields terrible results for smaller image sizes eg 512x512. This should default to 1 and users should be advised that they can increase this to improve performance.

File: invokeai/frontend/web/src/features/parameters/components/Parameters/Canvas/InfillAndScaling/ParamInfillPatchmatchDownscaleSize.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    dispatch(setInfillPatchmatchDownscaleSize(2));
+  }, [dispatch]);
+
+  return (
+    <IAISlider
+      isDisabled={infillMethod !== 'patchmatch'}
+      label={t('parameters.patchmatchDownScaleSize')}
+      min={1}
+      max={10}
+      sliderNumberInputProps={{ max: 10 }}
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/parameters/components/Parameters/Canvas/InfillAndScaling/ParamInfillPatchmatchDownscaleSize.tsx:48]: I don't think this prop is needed if the max is the same?

**Review by psychedelicious — CHANGES_REQUESTED** [review]
The changes here are difficult to compared to previous iterations; there are a lot of changes, so it's impossible to say which changes result in improvements. I did notice one thing, the patchmatch scale needs to be changed (see comment).

We are going to be shipping an update soon, so the model situation for LaMA needs to be sorted, else this PR will end up in 3.1.1 without model support.

File: invokeai/frontend/web/src/features/parameters/store/generationSlice.ts (hunk 1 above)
Comment (blessedcoolant) [reply invokeai/frontend/web/src/features/parameters/store/generationSlice.ts:95]: Done on both.

**Review by psychedelicious — APPROVED** [review]

**Review by lstein — APPROVED** [review]
LGTM

**Pull request merged**